### PR TITLE
only call res_init on Linux, to avoid threadsafety issues

### PR DESCRIPTION
**Pull request by oconnor663**
I finally feel like I have a better understanding of the platform issues involved in making this call. See the crash repro at https://github.com/rust-lang/rust/issues/43592 and the long screed in resinit_linux.go. What we're doing now is fairly conservative -- only calling `res_init` on non-Android Linux and _even still_ (though we expect most implementations there to be threadsafe) doing it inside a Mutex -- but we avoid dipping into really crazy glibc-runtime-symbol-detection CGO.

r? @maxtaco

### CODE REVIEW

File: rpc/resinit/resinit_linux.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+// though the network is up. The workaround for this bug is to call the
+// res_init() libc function to clear the cached configs, generally after a
+// failed network request.
+//
+// However, res_init() is documented *not threadsafe*. It really does crash on
+// OSX if you call it from multiple threads on e.g. macOS
+// (https://github.com/rust-lang/rust/issues/43592). Nor does putting a lock
+// around it totally let us off the hook. This is library code, after all, and
+// there's no telling what *other* libraries might be doing on parallel threads
+// in our process.
```
Comment (jzila) [inline rpc/resinit/resinit_linux.go]: Do all processes have their own resolver cache? That is, is a mutex enough? Can another process calling `res_init()` screw with us?

File: rpc/resinit/resinit.go [hunk 2]
```diff
@@ -2,10 +2,10 @@ package resinit
 
 import "net"
 
-// If we get a DNS error, it could be because glibc has cached an old version
-// of /etc/resolv.conf. The res_init() libc function busts that cache and keeps
-// us from getting stuck in a state where DNS requests keep failing even though
-// the network is up. This is similar to what the Rust standard library does:
+// Work around a glibc bug where cached confics from /etc/resolv.conf can cause
```
Comment (jzila) [inline rpc/resinit/resinit.go]: `s/confics/configs`?

File: rpc/resinit/resinit_linux.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+//       return funcPtr();
+//     }
+//   }
+//   */
+//   import "C"
+//
+// Instead of doing that...thing...we're going to make a simpler compromise:
+// 1) Only call res_init on (non-Android) Linux, which is almost certainly the
+//    only place we'll ever see glibc. Helpfully, the second mostly popular
+//    libc I know of on Linux, musl, also has a threadsafe/empty res_init().
```
Comment (jzila) [inline rpc/resinit/resinit_linux.go]: How about uClibc (for busybox)?

File: rpc/resinit/resinit_linux.go (hunk 1 above)
Comment (oconnor663) [inline rpc/resinit/resinit_linux.go]: The cache here is a glibc global variable, so I don't think other processes matter.

File: rpc/resinit/resinit.go (hunk 2 above)
Comment (oconnor663) [inline rpc/resinit/resinit.go]: Deliberate!

Comment (oconnor663) [inline rpc/resinit/resinit.go]: https://github.com/keybase/go-framed-msgpack-rpc/commit/00c2fe799455eb989b716289310b8406427785b1

File: rpc/resinit/resinit_linux.go (hunk 3 above)
Comment (oconnor663) [inline rpc/resinit/resinit_linux.go]: Thankfully, it does look like they take a lock: https://git.uclibc.org/uClibc/tree/libc/inet/resolv.c#n3546

Also to their credit, it sounds like they don't have the original glibc bug: https://git.uclibc.org/uClibc/tree/libc/inet/resolv.c#n533

```
 * We will read and analyze /etc/resolv.conf as needed before
 * we do a DNS request. This happens in __dns_lookup.
 * It is reread if its mtime is changed.
```

**Pull request merged**